### ISSUE TITLE
WebSocket Connection Interrupted on LabGraph Monitor

### ISSUE THREAD

**Issue by fzchriha** [facebookresearch/labgraph#69]
## 🐛 Bug

WebSocket connection is interrupted when trying to display the real-time messages on Labgraph Monitor (frontend)

## To Reproduce

1. python3 labgraph/examples/labgraph_monitor_example.py  
2. `cd extensions/prototypes/labgraph_monitor`
3. yarn start
4. Click on **RandomMessage** Node

```
- WS_SERVER_NODE: Raised an exception: RuntimeError("Task <Task pending name='Task-162949' coro=<WSAPIServerNode.ws_server_publisher() running at /Users/fatimazahra.chriha/Desktop/mlh/spring2022/labgraph/labgraph/websockets/ws_server/ws_api_node_server.py:125>> got Future <Future pending> attached to a different loop")
```

## Expected behavior

After a few seconds the real-time data freezes, check the terminal where the WebSocket is running it will display the error above

<img width="1000" alt="Screen Shot 2022-04-02 at 3 20 49 PM" src="https://user-images.githubusercontent.com/30349896/161398047-e25fdb9c-3ee5-4e51-9a95-511cd9a55e83.png">

## Environment

```

 - LabGraph Version:
 - OS: Mac
 - Python version: Python 3.8.13

**Comment by dtemir** [comment]
@fzchriha can you send a screenshot of the error you're getting? i'm trying understand the behavior from the gif, but it loops too fast I can't see what's written in the terminal

**Comment by fzchriha** [comment]
Hi @dtemir It's in the description above:
`- WS_SERVER_NODE: Raised an exception: RuntimeError("Task <Task pending name='Task-162949' coro=<WSAPIServerNode.ws_server_publisher() running at /Users/fatimazahra.chriha/Desktop/mlh/spring2022/labgraph/labgraph/websockets/ws_server/ws_api_node_server.py:125>> got Future <Future pending> attached to a different loop")`

**Comment by dtemir** [comment]
@fzchriha yeah I see! I was just wondering if there something useful in the Python traceback that shows what parts of the code the error breaks

**Comment by dtemir** [comment]
@fzchriha thank you! looking into that

**Comment by dtemir** [comment]
@fzchriha just to be sure, this has been resolved?

https://github.com/facebookresearch/labgraph/pull/63#issuecomment-1086724512

**Comment by fzchriha** [comment]
No it hasn't that comment was for realtime data sync with frontend

**Comment by bennaaym** [comment]
@fzchriha could you update PR #63 to use the last version of Labgraph Monitor and see if that solves the issue. I tried it and it looks like working fine

![demo](https://user-images.githubusercontent.com/68559207/162574598-c7107363-27bc-473b-a721-1943ff063d63.gif)

**Comment by fzchriha** [comment]
@bennaaym Hi Aymen I used the last version of labgraph monitor and for some reason when I try to connect using the latest changing @dtemir made (they work on my current PR) the status stays in **IS CONNECTING** and it never connects with the graph
<img width="317" alt="Screen Shot 2022-04-11 at 4 39 33 PM" src="https://user-images.githubusercontent.com/30349896/162828031-ad5d8c3e-3ddb-4d59-9820-8c9ae1b7011f.png">

**Comment by bennaaym** [comment]
@fzchriha, be sure that there is **.env.local** that has the following **key**:  REACT_APP_WS_API="ws://127.0.0.1:9000"

**Comment by fzchriha** [comment]
Oh okay

**Comment by fzchriha** [comment]
I have it there, but it still does not work

**Comment by bennaaym** [comment]
could you send another screenshot of both the server console and the frontend?

**Comment by fzchriha** [comment]
Frontend
<img width="1110" alt="Screen Shot 2022-04-11 at 4 59 47 PM" src="https://user-images.githubusercontent.com/30349896/162831226-9edd5ea5-d0eb-4221-ae49-e7f174980c02.png">

Backend
<img width="569" alt="Screen Shot 2022-04-11 at 5 00 38 PM" src="https://user-images.githubusercontent.com/30349896/162831322-7c5d0b6d-4968-4213-9a46-357b2e6ad297.png">

**Comment by bennaaym** [comment]
Could you try to restart the app, that's essential to pick up the changes in the  .env variables

**Comment by fzchriha** [comment]
Yes it won't work :(

**Comment by dtemir** [comment]
hey guys, just commenting to see if you want this issue closed as it now works in #71 ?

**Comment by fzchriha** [comment]
Hi @dtemir,unfortunately, it still does not work. I will discuss it with Jimmy again in our next meeting. In the meanwhile can you try testing (frontend + backend) on your Mac machine?

**Comment by dtemir** [comment]
@fzchriha sorry, I gave my mac to my sister for her school. But I just tested it on my Ubuntu 20.04 and the front-end + back-end seems to be working perfectly. Connected and disconnected a dozen of times without errors (except for back-end, which throws warnings whenever I disconnected). Do you think the bug is machine dependent then?

**Comment by fzchriha** [comment]
>  Do you think the bug is machine dependent then?

That's what I'm trying to test out, because Ayman tested it on windows & ubuntu and I'm the only one who has this issue & have a mac. So we are trying to weed out the possible causes. Otherwise, this error would be related to the websocket API config (labgraph internet code)

**Comment by dtemir** [comment]
> Hi @dtemir It's in the description above: `- WS_SERVER_NODE: Raised an exception: RuntimeError("Task <Task pending name='Task-162949' coro=<WSAPIServerNode.ws_server_publisher() running at /Users/fatimazahra.chriha/Desktop/mlh/spring2022/labgraph/labgraph/websockets/ws_server/ws_api_node_server.py:125>> got Future <Future pending> attached to a different loop")`
> 
> Here is the screenshot of the terminal <img alt="Screen Shot 2022-04-02 at 4 42 31 PM" width="858" src="https://user-images.githubusercontent.com/30349896/161400512-364e965e-6211-4aea-9c1e-b704ddaf57de.png">

Just to be sure, it throws the same error described above?

**Comment by fzchriha** [comment]
Yes

**Comment by garrettmflynn** [comment]
@fzchriha Happy to quickly test out the code on my Mac. 

Has the `labgraph/examples/labgraph_monitor_example.py` been renamed? I'm unable to find it in the repo.

**Comment by dtemir** [comment]
yeah it's now under `extensions/yam_support/labgraph_monitor/examples/labgraph_monitor_example.py`

**Comment by fzchriha** [comment]
@garrettmflynn thank you, Garret, can you please use the yaml_support folder @dtemir has in #64

**Comment by garrettmflynn** [comment]
Oh sure. So you don't want me to run the `extensions/yaml_support/labgraph_monitor/examples/labgraph_monitor_example.py` like in your original **To Reproduce** section?

I've got the frontend running—though it's still quite unclear to me how to get the server running at ws://127.0.0.1:9000 despite adding the required .env.local file.

**Comment by dtemir** [comment]
@garrettmflynn to run the server at `ws://127.0.0.1:9000` you need to run `python3 extensions/yaml_support/labgraph_monitor/examples/labgraph_monitor_example.py` and it requires labgraph

**Comment by dtemir** [comment]
> There's a persistent error on `homebrew install buck` and, subsequently, on `python3 setup.py install` when `buck` can't be found on the PATH. I think it's a permissions thing—but I've tried everything (including re-installing Homebrew) I can find on the internet to no avail.
> 
> Any tips or suggestions for installation?

Have you tried this approach by chance? https://github.com/facebookresearch/labgraph/blob/main/docs/installation/macos.md

I think we faced a similar issue with @fzchriha a while back, setting it up on Mac OS. Maybe that's the same problem here?

**Comment by garrettmflynn** [comment]
@dtemir Got a bit further. Any chance you know how to fix this error?

<img width="548" alt="Screen Shot 2022-04-14 at 4 06 51 PM" src="https://user-images.githubusercontent.com/46533749/163491596-d7a4cb45-86c8-4474-88cf-b0f9002a010d.png">

**Comment by dtemir** [comment]
@garrettmflynn just to be sure, what Python version are you using?

**Comment by garrettmflynn** [comment]
@dtemir Good question! I was on 3.9, though now have downgraded to 3.8.12. Here's the new error:
<img width="555" alt="Screen Shot 2022-04-15 at 9 17 04 AM" src="https://user-images.githubusercontent.com/46533749/163594916-dd4ed0b0-2d3c-423d-a01d-a396d4810aa2.png">

**Comment by dtemir** [comment]
@garrettmflynn thanks for trying it out with 3.8. Do you mind running `labgraph$ python -m labgraph.examples.simple_viz` just to make sure it's not just a python problem?

**Comment by garrettmflynn** [comment]
@dtemir After getting into this, I don't think I'll be able to get this test done. I'd assumed the process would be substantially easier—though it looks like my dev environment won't cooperate.

After I restarted my terminal, the `python -m labgraph.examples.simple_viz` errored out and the `labgraph/examples/labgraph_monitor_example.py` and its React counterpart in the Monitor errored after I clicked on the RandomMessage node inside the React UI. This was all on your branch, though, so I'd assume that's why it didn't work.

After this, I restarted the terminal again—and now neither of the files can be called again with [the same error](https://github.com/facebookresearch/labgraph/issues/69#issuecomment-1100207057). I've spent a few hours, on and off, trying to get it working—but no luck.

Any further suggestions would be appreciated! Otherwise I'll just have to bow out for now.

**Comment by dtemir** [comment]
no problem @garrettmflynn! thank you for taking your time
